### PR TITLE
Change default theme argument to file_html and components and update documentation

### PR DESCRIPTION
**Pull request by jeremy9959**
This PR addresses the rest of #6821.  It makes the following changes:
In the file ```bokeh/embed/standalone.py``` it changes the default value for the ```theme``` argument
in the functions ```components``` and ```file_html``` from ```FromCurdoc``` to ```None```. 

Typically these functions are used in situations where there is no ```curdoc```.  The value ```None``` means that if the function is passed a ```Document``` or a sequence of ```Model``` which constitute ```D.roots``` for some document ```D```, then the theme of that document will be applied; otherwise the default theme is applied.  Ultimately this stems from the behavior of the context manager ```OutputDocumentFor``` defined in ```embed.util```.

The default for the function ```json_item``` remains ```FromCurdoc``` on the theory that this function is most often used in server settings where ```curdoc``` makes sense.

In all cases the documentation of the keyword argument has been changed to more accurately describe the behavior of the various choices, following the logic of ```OutputDocumentFor```.

Finally, two example files are included.  In ```json_item_themed.py``` we show how to apply themes and how the interaction with ```curdoc``` works.  In ```file_html.py``` we recover the original sprint example from #6821.

### CODE REVIEW

**Comment by bryevdv** [comment]
> The default for the function json_item remains FromCurdoc on the theory that this function is most often used in server settings where curdoc makes sense.

@jeremy9959 I expect `json_items` is mostly used outside boken server applications, i.e. by Flask or other web apps or embedded in static pages.

**Comment by bryevdv** [comment]
> The default for the function json_item remains FromCurdoc on the theory that this function is most often used in server settings where curdoc makes sense.

@jeremy9959 I expect `json_items` is mostly used outside boken server applications, i.e. by Flask or other web apps or embedded in static pages.

**Comment by bryevdv** [comment]
> The default for the function json_item remains FromCurdoc on the theory that this function is most often used in server settings where curdoc makes sense.

@jeremy9959 I expect `json_items` is mostly used outside boken server applications, i.e. by Flask or other web apps or embedded in static pages.

**Comment by jeremy9959** [comment]
OK, then I can make the same change with json_items.

**Comment by bryevdv** [comment]
> The default for the function json_item remains FromCurdoc on the theory that this function is most often used in server settings where curdoc makes sense.

@jeremy9959 I actually expect `json_items` is mostly used outside boken server applications, i.e. by Flask or other web apps or embedded in static pages.

**Comment by bryevdv** [comment]
Sorry for (deleted) repeated comments, GH UI seems to have some hiccups this morning.

**Comment by jeremy9959** [comment]
I haven't worked with the ```typing``` module before and it caused trouble with the tests because ```theme``` has type ```ThemeLike = [Theme, Type[FromCurdoc]]```.  I added
None to that: ```ThemeLike = [None, Theme, Type[FromCurdoc]]``` but that may not work.  What's the right way to do it?

**Comment by jeremy9959** [comment]
I haven't worked with the ```typing``` module before and it caused trouble with the tests because ```theme``` has type ```ThemeLike = [Theme, Type[FromCurdoc]]```.  I added
None to that: ```ThemeLike = [None, Theme, Type[FromCurdoc]]``` but that may not work.  What's the right way to do it?

**Comment by bryevdv** [comment]
That's the right way to do it. As for GH I think they are having some sporadic backend issues right now

**Comment by bryevdv** [comment]
@jeremy9959 I will take a closer look tonight, if you have time, you could run the `examples/app/crossfilter` bokeh server app example just to verify that the theming there is still applied as expected (I feel confident it is but always good to check)

**Comment by jeremy9959** [comment]
```crossfilter``` looks fine
Screen Shot 2020-02-25 at 1 12 28 PM

**Comment by bryevdv** [comment]
I've restarted the test, however it appears there are some real layout baseline failures which would take some investigation. I'll look more closely tonight.

**Comment by jeremy9959** [comment]
I had missed this but ```file_html``` is used in three other places:
- ```io.saving``` 
- ```io.export``` 
- ```command.subcommands``` where it is used by the ```html``` command. 

In each case it's relying on the default option.  This is why, for example, the glyph page doesn't work.  I can imagine going into those three places and giving an explicit theme argument that reconstructs the original default... but maybe that is too messy.

**Comment by bryevdv** [comment]
@jeremy9959 I think I am also now remembering better where `FromCurdoc` came from. IIRC there were folks saving out the content generated inside a Bokeh app as standalone output (i.e. they wanted to generate HTML from inside app code. by calling `file_html`). I think this change might break that capability. We might want to take a short step back and actually enumerate the various use-cases, explicitly describe the desired behavior, and and make sure we can put them all under test.

**Comment by jeremy9959** [comment]
OK, fair enough.  I am perhaps over-enthusiastic here.

**Comment by bryevdv** [comment]
Enthusiasm is good! These code paths and requirements are just a bit convoluted (my fault) so I think taking a careful census is advisable :)

File: examples/embed/file_html.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
+from __future__ import print_function
```
Comment (mattpap) [inline examples/embed/file_html.py]: This isn't needed anymore (a Py2 thing).

Comment (jeremy9959) [reply examples/embed/file_html.py]: Thanks -

**Comment by jeremy9959** [comment]
In rethinking this one basic question is whether the smallest unit to which a ```Theme``` can be applied is a ```Document```.  The documentation seems to suggest this, it reads:

```
class Theme(object):
    ''' Provide new default values for Bokeh models.

    Bokeh Model properties all have some built-in default value. If a property
    has not been explicitly set (e.g. ``m.foo = 10``) then accessing the
    property with return the default value. It may be useful for users to be
    able to specify a different set of default values than the built-in
    default. The ``Theme`` class allows collections of custom default values
    to be easily applied to Bokeh documents.
```

The functions ```components``` and ```json_items``` in standalone do not accept a document as an argument; they want either a sequence of models or a model.  

The function ```file_html``` will accept a document, but it then replaces the document with its roots.  Finally all of these use the context manager ```OutputDocumentFor``` which takes a sequence of models and creates a document for them.

So what happens generally is:
- you have some model(s), which:
    - might not belong to a document at all
    - might belong to a single explicitly created document
    - might belong to an implicitly created document (from an app or in a notebook)
    - might belong to several different documents.
Then:

- if the models are exactly doc.roots for some doc, you just leave doc alone
- if none of the models have a document, you make one for them. 
- otherwise you pull the models out of their documents (if any) and put them temporarily into a new document. 

In the process you try to infer what theme to apply.  The logic (i.e. before any of the changes discussed above) says:
- If no theme is passed, use the curdoc().theme.  Note that sometimes this is completely irrelevant to the situation.
- If None is passed, and all models belong to the same document, apply that theme; otherwise apply the default theme.
- If a specific theme is passed, apply that one.

The one clear-cut situation where the current logic is wrong is:
- the models belong to an explicitly created document, not curdoc(), that has a theme applied,
and the call to ```file_html``` has no explicit theme specified.  This causes the curdoc() theme to be applied, instead of the documents theme.  Changing the default to ```None``` fixes this problem.

A second situation where ```file_html``` is called is in the logic of the io module.  There, it is accessed by the ```save``` function, and also by ```show``` when ```output_file``` is active.  In this situation, there is a ```curstate()``` and a ```curdoc()``` and the ```save``` function even takes a ```state``` as an argument.  In this case what seems reasonable to me is that the theme from  ```state``` argument's document, or else ```curdoc().theme``` by default, be used.  This means passing that information into ```_save_helper```.

I remain conflicted about how to handle ```components``` and ```json_items``` which apply to models, not documents, and so it is not clear how one should treat them as far as themes go.

There are two more places to look at : subcommands, and export.

**Comment by bryevdv** [comment]
> I remain conflicted about how to handle components and json_items which apply to models 

I'll have to respond more carefully after the release but wanted to mention quickly: all fo these functions really work on documents, and its probably best to think of them that way. The ones that accept models directly do that for the convenience of the user. But in those cases  as you've noted they either: use the document that the model is already a part of, or create a new temp document to use.

**Comment by jeremy9959** [comment]
Unfortunately neither json_items nor components work on documents.  

```python
from bokeh.embed.standalone import components, json_item
from bokeh.plotting import figure
from bokeh.document import Document

p = figure()
s = figure()

doc1 = Document()
doc1.add_root(p)
doc2 = Document()
doc2.add_root(s)

for func in [components, json_item]:

    try:
        _ = func(doc1)
    except ValueError as e:
        print(e,'{} failed with one doc'.format(func))

    try:
        _ = func([doc1, doc2])
    except ValueError as e:
        print(e,'{} failed with list of docs'.format(func))
 
    try:
        _ = func({'doc1':doc1,'doc2':doc2})
    except ValueError as e:
        print(e,'{} failed with dict of docs'.format(func))
```

yields
```
OutputDocumentFor expects a sequence of Models <function components at 0x7fd65d8a7200> failed with one doc
OutputDocumentFor expects a sequence of Models <function components at 0x7fd65d8a7200> failed with list of docs
OutputDocumentFor expects a sequence of Models <function components at 0x7fd65d8a7200> failed with dict of docs
OutputDocumentFor expects a sequence of Models <function json_item at 0x7fd65d8a7320> failed with one doc
OutputDocumentFor expects a sequence of Models <function json_item at 0x7fd65d8a7320> failed with list of docs
OutputDocumentFor expects a sequence of Models <function json_item at 0x7fd65d8a7320> failed with dict of docs
```

The logic of those functions doesn't handle docs correctly.  Not sure if this counts as a separate issue or all just part of the same phenomenon.

**Comment by bryevdv** [comment]
Sorry I did not mean to imply that those functions accept Documents as input (maybe they could, should, separate question) but rather that fundamentally they operate on Documents (that is what `OutputDocumentFor` is for, to get a Document, on way or another) and that it is useful to keep that frame in mind. This collection of functions grew up one at at time there is probably space to re-organize their implementations along cleaner lines.

**Comment by jeremy9959** [comment]
I had just decided that I couldn't understand the bokeh html command and why it didn't work with themes, but it looks like you dropped that command anyway. Am I right?

**Comment by bryevdv** [comment]
@jeremy9959 Yes, it has been entirely broken since 1.0 and not a single person noticed until this month. So, definitely not worth putting resources in to maintaining.

**Comment by jeremy9959** [comment]
OK, I just pushed my latest PR and I think I will leave it here for now.  I've done the testing that I can think of and it looks pretty good. I don't know much about how to formally set up unit tests but I'm interesting in learning.   Once you get a chance to look at it I am happy to do more but I would need some guidance.  In the end, I changed the default theme to ```None``` in ```json_item```, ```components```, and ```file_html```, and then modified the place in the ```io``` module where ```file_html``` was called to make sure that the theme of ```curdoc``` is explicitly applied.

**Comment by bryevdv** [comment]
@jeremy9959  awesome, let me get 2.0 the door this week and I will circle back to take a closer look and also make concrete suggestions about testing. Thanks for your contribution and patience!

**Comment by bryevdv** [comment]
Just noting for the future: the one unit test that is failing everywhere is due to the new required positional arg that was added to `_save_helpper`.

```
______________________________ test__save_helper _______________________________
259
260 mock_file_html = <MagicMock name='file_html' id='140369359355856'>
261 mock_io_open = <MagicMock name='open' id='140369359479416'>
262 
263    @patch('io.open')
264    @patch('bokeh.embed.file_html')
265    def test__save_helper(mock_file_html, mock_io_open) -> None:
266        obj = Plot()
267        filename, resources, title = bis._get_save_args(curstate(), "filename", "resources", "title")
268    
269 >       bis._save_helper(obj, filename, resources, title, None)
270 E       TypeError: _save_helper() missing 1 required positional argument: 'theme'
271 
272 tests/unit/bokeh/io/test_saving.py:104: TypeError
```

possible corrections include:

* making theme an optional parameter with a default
* updating the test to pass a theme value

**Comment by bryevdv** [comment]
@jeremy9959 do you have thoughts on the comment above about the API change above? 

Also, there are two failures in the examples tests:

>/home/runner/work/bokeh/bokeh/tests/test_examples.py:203: AssertionError: examples/embed/file_html.py doesn't have a baseline
/home/runner/work/bokeh/bokeh/tests/test_examples.py:51: AssertionError: examples/embed/json_item_themed.py timed out

The first we can help with generating the necessary baseline file. Im not sure what's up with the secon (timeout) one

**Comment by jeremy9959** [comment]
Do you have a standard about whether "private" methods should allow default arguments? If not,
I think the simplest is to make save_helper have a default argument.  I can fix that.
I am not sure about the other error but I will take a look at it.

**Comment by bryevdv** [comment]
@jeremy9959 we guidelines there,  a default param is fine with me

**Comment by jeremy9959** [comment]
I added a default argument ```theme=None``` to ```_save_helper``` but I still get a unit test failure:
```
_____________________________________________________ test__save_helper ______________________________________________________

mock_file_html = <MagicMock name='file_html' id='140646747025520'>
mock_io_open = <MagicMock name='open' id='140646747037456'>

    @patch('io.open')
    @patch('bokeh.embed.file_html')
    def test__save_helper(mock_file_html, mock_io_open) -> None:
        obj = Plot()
        filename, resources, title = bis._get_save_args(curstate(), "filename", "resources", "title")

        bis._save_helper(obj, filename, resources, title, None)

        assert mock_file_html.call_count == 1
        assert mock_file_html.call_args[0] == (obj, resources)
>       assert mock_file_html.call_args[1] == dict(title="title", template=None)
E       AssertionError: assert {'template': ...tle': 'title'} == {'template': ...tle': 'title'}
E         Omitting 2 identical items, use -vv to show
E         Left contains 1 more item:
E         {'theme': None}
E         Use -v to get the full diff

test_saving.py:108: AssertionError
```
I am not entirely sure of how this is all set up so before modifying the test or doing something else I could use some advice.

**Comment by bryevdv** [comment]
@jeremy9959 I think you can add `theme=None` to the dict in this line in the test:
```
assert mock_file_html.call_args[1] == dict(title="title", template=None)
```
TBH these existing mock tests here are not great but for now at least better than nothing

**Comment by jeremy9959** [comment]
I changed the file ```examples.yaml``` in the tests directory so that ```json_item_themed.py``` was listed as "skip" along with ```json_item.py``` which was the original example without a theme.  That cured one test failure.  All that's left is the baseline.  Happy to work on that, too, if you can point me to what those are.

**Comment by bryevdv** [comment]
@mattpap can you help with the baseline? I don't have a linux VM handy to use to generate it

**Comment by jeremy9959** [comment]
@mattpap you worked your baseline magic on my other PR #9755 can you help me with this one? I'd like to tackle something else if this is ok.

**Comment by bryevdv** [comment]
@jeremy9959 apologies this has languished somewhat. I think if you rebase or resubmit this on current master, then CI might now show the expected baseline in the failure (which can then just be added to get things passing). Can you rebase or resubmit?

I'm less sure how the integration images are generates. @mattpap are there docs somewere?

**Comment by jeremy9959** [comment]
The failure occurs with `examples/embed/file_html.py` which makes the sprinter graph.  The graph now looks correct (i.e. the theme is applied).  Not sure what to do now....
Screenshot from 2020-05-22 12-10-37

**Comment by bryevdv** [comment]
@jeremy9959 I guess the file has to be present at all to get the diff. Can you check in an empty file:
```
tests/baselines/examples/embed/file_html
```
@mattpap def suggest these also spit out the expected content when the file is missing

**Comment by mattpap** [comment]
> def suggest these also spit out the expected content when the file is missing

It's supposed to do that, but apparently doesn't. A workaround may be to commit an empty baseline, then it should definitively show a diff.

**Comment by jeremy9959** [comment]
I have no idea why I got some JS test failures, but here's the message related to file_html:
```
[FAIL] diff --git a/tests/baselines/examples/embed/file_html b/tests/baselines/examples/embed/file_html
[FAIL] index e69de29..08f8635 100644
[FAIL] --- a/tests/baselines/examples/embed/file_html
[FAIL] +++ b/tests/baselines/examples/embed/file_html
[FAIL] @@ -0,0 +1,4 @@
[FAIL] +Plot bbox=[0, 0, 1000, 600]
[FAIL] +  LinearAxis bbox=[5, 554, 953, 46]
[FAIL] +  LinearAxis bbox=[958, 28, 42, 526]
[FAIL] +  Title bbox=[5, 0, 953, 28]
FAILED
```

**Comment by bryevdv** [comment]
@jeremy9959 that diff gives the expected contents of the `file_html` baseline. To save some iteration cycles, I've just pushed that content to the file here from the GH web UI. (Fingers crossed will go green now!)

**Comment by bryevdv** [comment]
OK well.. 

```
[FAIL] -Plot bbox=[0, 0, 1000, 600]
[FAIL] -  LinearAxis bbox=[5, 554, 953, 46]
[FAIL] -  LinearAxis bbox=[958, 28, 42, 526]
[FAIL] -  Title bbox=[5, 0, 953, 28]
[FAIL] +Plot bbox=[0, 0, 1000, 600]
[FAIL] +  LinearAxis bbox=[5, 554, 953, 46]
[FAIL] +  LinearAxis bbox=[958, 28, 42, 526]
[FAIL] +  Title bbox=[5, 0, 953, 28]
```
I would have thought maybe a tabs/space issue but even the line with no leading space fail to match. @mattpap ideas?

**Comment by bryevdv** [comment]
All I can think of is maybe a CF/LF issue from having edited in the GH web UI. @jeremy9959 can you pull the update and the open and re-save in your own local editor and see if that actually records any change?

**Comment by mattpap** [comment]
> All I can think of is maybe a CF/LF (...)

Definitively.

**Comment by bryevdv** [comment]
The thing is is that I've definitely successfully made small edits to baselines in the GH web UI directly before. It's much less work than editing a branch on a fork. Don't our custom lint tools verify line endings on all files? 

Regardless, how can we make this more resilient/simple to update? Perhaps we can create a docker image just for running examples tests so that that examples tests are actualy runnable/reproducible outside CI by anyone, and that also has easy to run tools for automatically updating/committing baselines.

**Comment by mattpap** [comment]
> The thing is is that I've definitely successfully made small edits to baselines in the GH web UI directly before. It's much less work than editing a branch on a fork.

Sure. It's a pity that GitHub doesn't allow to at least configure line endings, if not respect the repository default.

> Codebase tests pass, I thought our custom lint tools verify line endings on all files?

I'm pretty sure they don't consider baselines.

> Regardless, how can we make this more resilient/simple to update? 

The standard is not to mix different line endings. We need to update our code quality tests to cover all files to point out this kind of issues in future.

**Comment by bryevdv** [comment]
> The standard is not to mix different line endings. We need to update our code quality tests to cover all files to point out this kind of issues in future.

Agree 100% but I am also interested in lowering the bar to accessibility so that anyone can have a reasonable chance at running and fixing things locally with minimal effort, which is definitely not the case now.

**Comment by jeremy9959** [comment]
@bryevdv did I screw something up that you had already fixed?  I got your first note (asking me to update file_html) so I did that, then logged on and saw maybe you'd already fixed it?  And now I am getting other weird errors.  Sigh.

**Comment by bryevdv** [comment]
@jeremy9959 Sorry for not communicating more clearly, @mattpap did push a change to try and get the last obstacle cleared  I'm not sure if your latest push "undid" anything, we will have to wait for the examples tests to pass. If they pass then I will merge this, if not we can try rolling back to https://github.com/bokeh/bokeh/pull/9729/commits/52169fa155970144fb3d97802030457fbc712a2c

**Comment by bryevdv** [comment]
@mattpap it looks like we are not using the right version of BokehJS in the tests:
> [JS] {warning} file:///home/runner/miniconda3/lib/python3.7/site-packages/bokeh/server/static/js/bokeh.js:1105:34 [bokeh] Library versions: JS (2.0.2) / Python (145fe2c)

I had noticed something similar in a different context in my WIP ci refactor PR but it seems like perhaps there is a more general problem. I'll make a note in the PR about this. 

@jeremy9959 the unit test failures are unrelated network/GH infra problems. It looks like examples tests will go green in which case I will go ahead and merge. Thank you for all your patience in getting this pushed through!

**Comment by jeremy9959** [comment]
This has been fun!  Thank YOU for your patience!  I learned a lot from this.

**Comment by bryevdv** [comment]
OK, examples DID go green. But looking one last time I think there are some unrelated changes to two package lock files that snuck in:

* bokehjs/package-lock.json
* examples/custom/latex_extension_full/latex_label/package-lock.json

@jeremy9959 one last ask, can you remove any changes to these two files from this PR?

**Comment by bryevdv** [comment]
Thanks @jeremy9959 !

**Pull request merged**